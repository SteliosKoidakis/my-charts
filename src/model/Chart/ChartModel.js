
class ChartModel {
  constructor(type) {
    this.tablet = 0;
    this.smartphone = 0;
    this.total = 0;
    this.type = type;
    this.total = 0;
  }
  get tabletPercentage() {
    return this.tablet / this.total;
  }
  get smartphonePercentage() {
    return this.smartphone / this.total;
  }
  async getChartData () {
    // todo: check for alias use
    const items = await import(`../../../static/${this.type}.js`);
    this._updateData(items.default);
  }
  _updateData(items ) {
    this.tablet = items.tablet;
    this.smartphone = items.smartphone;
    this.total = items.tablet + items.smartphone;
  }
}

export default ChartModel;
