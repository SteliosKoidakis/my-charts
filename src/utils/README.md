# Utilities
Global utilities functions

## Usage
```javascript

import { formatNumber } from '~/utils';
```



###  formatNumber()

Format number with dots.


```javascript
// parameter number , type number
formatNumber(123345) 

// returns string
--> '123.345'

```

###  renderStyles()
Render styles to shadow dom

```javascript
// parameter shadow element
// parameter styles to be rendered
renderStyles(shadow, styles) 


```
###  renderDonutChart()
Render donut chart to an html element of shadow dom

```javascript
renderDonutChart({
   smartphone: {
     value: '',   // Value of smartphone data.
     color: ''     // Chart color of smartphone data.
   },
    tablet: {
      value: '', // Value of tablet data.
      color: '' // Chart color of tablet data.
    },
    element:{
      shadow: '', // Shadow dom.
      class: '', // Class to render the chart.
    },
}) 


```
###  renderLineChart()
Render line chart to an html element of shadow dom

```javascript
renderLineChart({
    color: '', // Color of line.
    data: [], // Information to generate the line of the chart.
    element:{
      backgroundClass:'' // Class for attach styles on line chart area.
      shadow: '', // Shadow dom.
      class: '', // Class to render the chart.
    },
}) 

```
 