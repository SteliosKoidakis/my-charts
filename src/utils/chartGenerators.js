import { select } from 'd3-selection';
import { scaleOrdinal, scaleTime, scaleLinear } from 'd3-scale';
import { arc, pie, line } from 'd3-shape';
import { max, extent } from 'd3-array';

const renderDonutChart = ({
  smartphone,
  tablet,
  element,
}) => {
  const data = [{
    name: 'smartphone',
    value: smartphone.smartphonePercentage,
  },
  {
    name: 'tablet',
    value: tablet.tabletPercentage,
  },
  ];
  const width = 150;
  const height = 150;
  const thickness = 8;
  const radius = Math.min(width, height) / 2;
  const color = scaleOrdinal([smartphone.smartphoneColor, tablet.tabletColor]);

  const svg = select(element.shadow)
    .select(`.${element.class}`)
    .append('svg')
    .attr('class', 'pie')
    .attr('width', width)
    .attr('height', height);

  const g = svg
    .append('g')
    .attr('transform', `translate(${width / 2},${height / 2})`);

  const arcItem = arc()
    .innerRadius(radius - thickness)
    .outerRadius(radius);

  const pieItem = pie()
    .value((d) => d.value)
    .sort(null);

  const path = g // eslint-disable-line no-unused-vars
    .selectAll('path')
    .data(pieItem(data))
    .enter()
    .append('g')
    .append('path')
    .attr('d', arcItem)
    .attr('fill', (d, i) => color(i));
};

const renderLinarChart = ({
  data,
  color,
  element,
}) => {
  const margin = {
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
  };

  const svg = select(element.shadow)
    .select(`.${element.class}`)
    .append('svg')
    .attr('width', 100 + margin.left + margin.right)
    .attr('height', 100 + margin.top + margin.bottom)
    .append('g')
    .attr('transform', `translate(${margin.left},${margin.top})`);

  const x = scaleTime()
    .domain(extent(data, (value) => value))
    .range([0, 100]);
  svg.append('g');

  const y = scaleLinear()
    .domain([0, max(data, (value) => +value)])
    .range([100, 0]);
  svg.append('g');

  svg
    .append('path')
    .datum(data)
    .attr('fill', 'none')
    .attr('stroke', color)
    .attr('stroke-width', 1.5)
    .attr(
      'd',
      line()
        .x((value, index) => x(index))
        .y((value) => y(value)),
    );
};

export {
  renderDonutChart,
  renderLinarChart
};
