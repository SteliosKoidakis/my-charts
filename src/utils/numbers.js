const formatNumber = (number) => {
  if (typeof number !== 'number') return number;

  return number.toString().replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1.');
};

export { formatNumber };
