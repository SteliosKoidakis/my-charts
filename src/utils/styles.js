const renderStyles = (shadow, styles) => {
  const style = document.createElement('style');
  style.textContent = styles;
  shadow.appendChild(style);
};

export { renderStyles };
