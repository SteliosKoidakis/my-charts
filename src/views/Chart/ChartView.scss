.ChartView {
  text-align: center;
  padding-top: 100px;

  &__item {
    display: inline-block;
    padding: 0 20px 10px 0;
    border-bottom: grey solid 0.5px;
  }

  @media (max-width: 900px) {
    width: 300px;
    overflow: hidden;
    margin: 0 auto;
    
    &__slides {
      display: flex;
      overflow-x: auto;
      scroll-snap-type: x mandatory;
      scroll-behavior: smooth;
    }

    &__slides::-webkit-scrollbar {
      width: 8px;
      height: 8px;
    }

    &__slides::-webkit-scrollbar-thumb {
      background: black;
    }

    &__item {
      position: relative;
      display: flex;
      justify-content: center;
      align-items: center;
      flex-shrink: 0;
      margin-right: 20px;
    }
  }
}