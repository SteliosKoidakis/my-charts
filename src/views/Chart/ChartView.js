import ChartModel from '~/model/Chart/ChartModel';
import { attachStyles } from '~/utils';
import ChartViewStyles from './ChartView.scss';

class ChartView extends HTMLElement {
  constructor() {
    super();
    this.shadow = this.attachShadow({
      mode: 'open',
    });

    this.revenewChart = new ChartModel('revenewChart');
    this.impresionsChart = new ChartModel('impresionsChart');
    this.visitsChart = new ChartModel('visitsChart');
  }

  async connectedCallback() {
    await this.getChartsInformation();
    this.renderTemplate();
    attachStyles(this.shadow, ChartViewStyles);
  }

  async getChartsInformation() {
    await this.revenewChart.getChartData();
    await this.impresionsChart.getChartData();
    await this.visitsChart.getChartData();
  }

  renderTemplate() {
    const template = `
      <div div class="ChartView">
        <div class="ChartView__item">
          ${this.revenewChart.total}
          ${this.revenewChart.tabletPercentage}
          ${this.revenewChart.smartphonePercentage}
          <donat-chart-component
            data=${this.revenewChart.data}
            smartphonePercentage=${this.revenewChart.smartphonePercentage}
            tabletPercentage=${this.revenewChart.tabletPercentage}
            smartphoneColor="#38AF1E"
            title="REVENUE"
            tabletColor="#44DF22"
            total=${this.revenewChart.total}
          />
        </div>
        <div class="ChartView__item">
          ${this.impresionsChart.total}
          ${this.impresionsChart.tabletPercentage}
          ${this.impresionsChart.smartphonePercentage}
          <donat-chart-component
            data=${this.impresionsChart.data}
            smartphonePercentage=${this.impresionsChart.smartphonePercentage}
            tabletPercentage=${this.impresionsChart.tabletPercentage}
            smartphoneColor="#2945BB"
            title="IMPRESIONS"
            tabletColor="#4DD4DD"
            total=${this.impresionsChart.total}
          />
        </div>
        <div class="ChartView__item">
          ${this.visitsChart.total}
          ${this.visitsChart.tabletPercentage}
          ${this.visitsChart.smartphonePercentage}
          <donat-chart-component
            data=${this.visitsChart.data}
            smartphonePercentage=${this.visitsChart.smartphonePercentage}
            tabletPercentage=${this.visitsChart.tabletPercentage}
            smartphoneColor="#E55039"
            title="VISITS"
            tabletColor="#E5C839"
            total=${this.visitsChart.total}
          />
        </div>
      </div>
      `;

    this.shadow.innerHTML = template;
  }
}

window.customElements.define('chart-view', ChartView);

export default ChartView;
