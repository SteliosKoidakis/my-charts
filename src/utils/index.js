import { renderDonutChart, renderLinarChart } from './chartGenerators';
import { renderStyles } from './styles';

export {
  renderStyles,
  renderDonutChart,
  renderLinarChart
};
