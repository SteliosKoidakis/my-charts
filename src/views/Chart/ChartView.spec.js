import ChartView from './ChartView';
import ChartModel from '~/model/Chart/ChartModel';
import { renderStyles } from '~/utils';

jest.mock('~/utils', () => ({
  renderStyles: jest.fn(),
}));

describe('Given component ChartView', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = new ChartView();
  });
  describe('When class is constructed', () => {
    it('Then revenewChart should be an instance of ChartModel("revenewChart")', () => {
      expect(wrapper.revenewChart).toMatchObject(new ChartModel('revenewChart'));
    });
    it('And impresionsChart should be an instance of ChartModel("impresionsChart")', () => {
      expect(wrapper.impresionsChart).toMatchObject(new ChartModel('impresionsChart'));
    });
    it('And visitsChart should be an instance of ChartModel("visitsChart")', () => {
      expect(wrapper.visitsChart).toMatchObject(new ChartModel('visitsChart'));
    });
  });
  describe('Given method connectedCallback', () => {
    describe('When is called', () => {
      it('Then it should trigger getChartsInformation, renderTemplate and renderStyles method', async () => {
        wrapper.getChartsInformation = jest.fn();
        wrapper.renderTemplate = jest.fn();
        await wrapper.connectedCallback();
        expect(wrapper.getChartsInformation).toHaveBeenCalled();
        expect(wrapper.renderTemplate).toHaveBeenCalled();
        expect(renderStyles).toHaveBeenCalled();
      });
    });
  });
  describe('Given method getChartsInformation', () => {
    describe('When is called', () => {
      it('Then it should trigger revenewChart, impresionsChart and visitsChart getChartData method', async () => {
        wrapper.revenewChart.getChartData = jest.fn();
        wrapper.impresionsChart.getChartData = jest.fn();
        wrapper.visitsChart.getChartData = jest.fn();
        await wrapper.getChartsInformation();
        expect(wrapper.revenewChart.getChartData).toHaveBeenCalled();
        expect(wrapper.impresionsChart.getChartData).toHaveBeenCalled();
        expect(wrapper.visitsChart.getChartData).toHaveBeenCalled();
      });
    });
  });
  describe('Given method renderTemplate', () => {
    describe('When is called', () => {
      it('Then it should render the correct template', () => {
        wrapper.renderTemplate();
        expect(wrapper.shadow.innerHTML).toMatchSnapshot();
      });
    });
  });
});
