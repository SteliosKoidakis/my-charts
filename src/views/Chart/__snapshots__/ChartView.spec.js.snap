// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given component ChartView Given method renderTemplate When is called Then it should render the correct template 1`] = `
"
    <div class=\\"ChartView\\">
      
      <div class=\\"ChartView__navigation\\">
        <a href=\\"#slide-1\\"></a>
        <a href=\\"#slide-2\\"></a>
        <a href=\\"#slide-3\\"></a>
      </div>
      <div class=\\"ChartView__slides\\">
        <div class=\\"ChartView__item\\" id=\\"slide-1\\">
          <donat-chart-component data=\\"smartphonePercentage=NaN\\" tabletpercentage=\\"NaN\\" smartphone=\\"0\\" tablet=\\"0\\" currency=\\"â‚¬\\" smartphonecolor=\\"#38AF1E\\" title=\\"REVENUE\\" tabletcolor=\\"#44DF22\\" total=\\"0\\">
        </donat-chart-component></div>
        <div class=\\"ChartView__item\\" id=\\"slide-2\\">
          <donat-chart-component data=\\"smartphonePercentage=NaN\\" tabletpercentage=\\"NaN\\" smartphone=\\"0\\" tablet=\\"0\\" smartphonecolor=\\"#2945BB\\" title=\\"IMPRESIONS\\" tabletcolor=\\"#4DD4DD\\" total=\\"0\\">
        </donat-chart-component></div>
        <div class=\\"ChartView__item\\" id=\\"slide-2\\">
          <donat-chart-component data=\\"smartphonePercentage=NaN\\" tabletpercentage=\\"NaN\\" smartphone=\\"0\\" tablet=\\"0\\" smartphonecolor=\\"#E55039\\" title=\\"VISITS\\" tabletcolor=\\"#E5C839\\" total=\\"0\\">
        </donat-chart-component></div>
      </div>
    </div>
      "
`;
